### PR TITLE
[IMP] l10n_latam_invoice_document: remove fields and compute methods

### PR DESCRIPTION
**Pull request by bruno-zanotti**
**Description of the issue/feature this PR addresses:**
Fields removed from l10n_latam:
 * l10n_latam_amount_untaxed
 * l10n_latam_tax_ids
 * l10n_latam_price_unit
 * l10n_latam_price_subtotal
 * l10n_latam_price_net

Methods removed from l10n_latam:
 * compute_l10n_latam_prices_and_taxes
 * compute_l10n_latam_amount_and_taxes
 * _compute_invoice_taxes_by_group

Move the removed methods to l10n_ar and l10n_cl and now it returns
a dictionary with the removed fields as the keys.

Also, adapt the invoices reports to get the required values from the new methods for each localization.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by zaoral — APPROVED** [review]
+1. 
Excellent @Bruno-Zanotti! 
remember to squash before merge

File: addons/l10n_latam_invoice_document/models/account_move.py [hunk 1]
```diff
@@ -128,30 +126,27 @@ def _get_starting_sequence(self):
...
         return super(AccountMove, self)._get_starting_sequence()
 
-    def _compute_l10n_latam_amount_and_taxes(self):
-        recs_invoice = self.filtered(lambda x: x.is_invoice())
-        for invoice in recs_invoice:
-            tax_lines = invoice.line_ids.filtered('tax_line_id')
-            currencies = invoice.line_ids.filtered(lambda x: x.currency_id == invoice.currency_id).mapped('currency_id')
-            included_taxes = invoice.l10n_latam_document_type_id and \
-                invoice.l10n_latam_document_type_id._filter_taxes_included(tax_lines.mapped('tax_line_id'))
+    def _l10n_latam_amount_and_taxes(self):
```
Comment (celm1990) [inline addons/l10n_latam_invoice_document/models/account_move.py:129]: Please can you add docstring for explain purpose this function

File: addons/l10n_ar/views/report_invoice.xml [hunk 2]
```diff
@@ -114,32 +114,35 @@
...
-            <attribute name="t-field">line.l10n_latam_price_unit</attribute>
+            <attribute name="t-value">latam_values['l10n_latam_price_unit']</attribute>
         </xpath>
         <xpath expr="//span[@id='line_tax_ids']" position="attributes">
-            <attribute name="t-esc">', '.join(map(lambda x: (x.description or x.name), line.l10n_latam_tax_ids))</attribute>
+            <attribute name="t-esc">', '.join(map(lambda x: (x.description or x.name), latam_values['l10n_latam_tax_ids']))</attribute>
         </xpath>
         <t t-set="current_subtotal" t-value="current_subtotal + line.price_subtotal" position="attributes">
-            <attribute name="t-value">current_subtotal + line.l10n_latam_price_subtotal</attribute>
+            <attribute name="t-value">current_subtotal + line._l10n_latam_prices_and_taxes()['l10n_latam_price_subtotal']</attribute>
```
Comment (Bruno-Zanotti) [inline addons/l10n_ar/views/report_invoice.xml:127]:
```suggestion
            <attribute name="t-value">current_subtotal + l10n_latam_price_subtotal['l10n_latam_price_subtotal']</attribute>
```

**Comment by jco-odoo** [comment]
@oco-odoo Is this overlapping with what you are doing in the withholding branch?

**Comment by vbe-odoo** [comment]
@Danisan Can you take a look at it.
It may affect CL, you can ping @Bruno-Zanotti for questions.
Thanks

File: addons/l10n_ar/models/account_move.py [hunk 3]
```diff
@@ -293,3 +293,73 @@ def _get_name_invoice_report(self):
...
+            else:
+                included_invoice_taxes = tax_lines.filtered(lambda x: x.tax_line_id in included_taxes)
+                not_included_invoice_taxes = tax_lines - included_invoice_taxes
+                sign = -1 if self.is_inbound() else 1
+                l10n_latam_amount_untaxed = self.amount_untaxed + sign * sum(included_invoice_taxes.mapped('balance'))
+            res['l10n_latam_amount_untaxed'] = l10n_latam_amount_untaxed
+            res['l10n_latam_tax_ids'] = not_included_invoice_taxes
+        else:
+            res['l10n_latam_amount_untaxed'] = False
+            res['l10n_latam_tax_ids'] = [(5, 0)]
```
Comment (jco-odoo) [inline addons/l10n_ar/models/account_move.py:316]: This is the worst name chosen: l10n_latam_tax_ids while these are account.move.lines

**Comment by celm1990** [comment]
Hi guys, target can be V15 instead master????

File: addons/l10n_ar/models/account_move.py [hunk 4]
```diff
@@ -293,3 +292,15 @@ def _get_name_invoice_report(self):
         if self.l10n_latam_use_documents and self.company_id.account_fiscal_country_id.code == 'AR':
             return 'l10n_ar.report_invoice_document'
         return super()._get_name_invoice_report()
+
+    def _prepare_tax_lines_data_for_totals_from_invoice(self, tax_line_id_filter=None, tax_ids_filter=None):
+        if self.company_id.account_fiscal_country_id.code == 'AR' and self._l10n_ar_include_vat() and (
+                'commit_assetsbundle' in self.env.context or not self.env.context.get('params', {}).get('view_type') == 'form'):
```
Comment (jco-odoo) [inline addons/l10n_ar/models/account_move.py:298]: We do not need this anymore.

File: addons/account/models/account_move.py [hunk 5]
```diff
@@ -1640,10 +1640,10 @@ def _compute_tax_totals_json(self):
                 move.tax_totals_json = None
                 continue
 
-            tax_lines_data = move._prepare_tax_lines_data_for_totals_from_invoice()
+            tax_lines_data, amount_untaxed = move._prepare_tax_lines_data_for_totals_from_invoice()
 
             move.tax_totals_json = json.dumps({
-                **self._get_tax_totals(move.partner_id, tax_lines_data, move.amount_total, move.amount_untaxed, move.currency_id),
+                **self._get_tax_totals(move.partner_id, tax_lines_data, move.amount_total, amount_untaxed, move.currency_id),
```
Comment (jco-odoo) [inline addons/account/models/account_move.py:1646]: Why do you need to hack into the account module here?  @jjscarafia @Bruno-Zanotti

Comment (jco-odoo) [reply addons/account/models/account_move.py:1646]: What could be wrong with amount_untaxed on the move? @oco-odoo

File: addons/l10n_cl/models/account_move.py [hunk 6]
```diff
@@ -124,3 +123,15 @@ def _get_name_invoice_report(self):
         if self.l10n_latam_use_documents and self.company_id.account_fiscal_country_id.code == 'CL':
             return 'l10n_cl.report_invoice_document'
         return super()._get_name_invoice_report()
+
+    def _prepare_tax_lines_data_for_totals_from_invoice(self, tax_line_id_filter=None, tax_ids_filter=None):
+        if self.company_id.account_fiscal_country_id.code == 'CL' and self._l10n_cl_include_sii() and (
```
Comment (jco-odoo) [inline addons/l10n_cl/models/account_move.py:128]: same here: do not hack the existing methods, but use another one.

**Review by jco-odoo — CHANGES_REQUESTED** [review]
I agree with @celm1990 that this should be targetting saas-14.5.  And please avoid hacking the Odoo standard or let's discuss about it really carefully.

File: addons/l10n_ar/models/account_move.py (hunk 3 above)
Comment (jjscarafia) [reply addons/l10n_ar/models/account_move.py]: Changed

**Review by celm1990 — APPROVED** [review]
LGTM

**Comment by jco-odoo** [comment]
@Bruno-Zanotti @jjscarafia PR is now needed to 15.0 instead of master.  It should not be difficult to cherry-pick what we have here.

**Comment by jco-odoo** [comment]
@Bruno-Zanotti also looks ok on my side.  Just need to squash and pr to 15.0 instead.

**Comment by vbe-odoo** [comment]
@jco-odoo @jjscarafia @Bruno-Zanotti @Danisan 
I run some FUN tests and LGTM (I've also tried CL as well).
Thanks.

**Comment by Bruno-Zanotti** [comment]
@jco-odoo @vbe-odoo @jjscarafia new pr on V15: https://github.com/odoo/odoo/pull/77961

**Pull request closed without merging**